Objects/dictobject.c

```diff
@@ -6820,53 +7000,119 @@ PyObject_VisitManagedDict(PyObject *obj, visitproc visit, void *arg)
-            PyDictValues *values = _PyObject_InlineValues(obj);
-            if (values->valid) {
-                for (Py_ssize_t i = 0; i < values->capacity; i++) {
-                    Py_CLEAR(values->values[i]);
-                }
-                values->valid = 0;
+        PyDictObject *dict = _PyObject_GetManagedDict(obj);
+        if (dict == NULL) {
+#ifdef Py_GIL_DISABLED
+            Py_BEGIN_CRITICAL_SECTION(obj);
+
+            dict = _PyObject_ManagedDictPointer(obj)->dict;
+            if (dict == NULL) {
+                set_dict_inline_values(obj, (PyDictObject *)new_dict);
             }
+
+            Py_END_CRITICAL_SECTION();
+
+            if (dict == NULL) {
+                return;
+            }
+#else
+            set_dict_inline_values(obj, (PyDictObject *)new_dict);
+            return;
+#endif
         }
+
+        Py_BEGIN_CRITICAL_SECTION2(dict, obj);
+
+#ifdef Py_DEBUG
+        // If the dict in the object has been replaced between when we got
+        // the dict and unlocked the objects then it's definitely no longer
+        // inline and there's no need to detach it, we can just replace it.
+        // The call to _PyDict_DetachFromObject will be a nop.
+        PyDictObject *cur_dict = _PyObject_ManagedDictPointer(obj)->dict;
+        assert(cur_dict == dict ||
+               (cur_dict->ma_values != _PyObject_InlineValues(obj) &&
+                dict->ma_values != _PyObject_InlineValues(obj) &&
+                !_PyObject_InlineValues(obj)->valid));
+#endif
+
+        FT_ATOMIC_STORE_PTR(_PyObject_ManagedDictPointer(obj)->dict,
+                            (PyDictObject *)Py_XNewRef(new_dict));
+
+        _PyDict_DetachFromObject(dict, obj);
+
+        Py_END_CRITICAL_SECTION2();
+
+        Py_DECREF(dict);
     }
     else {
-        Py_CLEAR(_PyObject_ManagedDictPointer(obj)->dict);
+        PyDictObject *dict;
+
+        Py_BEGIN_CRITICAL_SECTION(obj);
+
+        dict = _PyObject_ManagedDictPointer(obj)->dict;
+
+        FT_ATOMIC_STORE_PTR(_PyObject_ManagedDictPointer(obj)->dict,
+                            (PyDictObject *)Py_XNewRef(new_dict));
+
+        Py_END_CRITICAL_SECTION();
+
+        Py_XDECREF(dict);
     }
     assert(_PyObject_InlineValuesConsistencyCheck(obj));
 }
 
+void
+PyObject_ClearManagedDict(PyObject *obj)
+{
+    _PyObject_SetManagedDict(obj, NULL);
+}
+
 int
 _PyDict_DetachFromObject(PyDictObject *mp, PyObject *obj)
 {
-    assert(_PyObject_ManagedDictPointer(obj)->dict == mp);
-    assert(_PyObject_InlineValuesConsistencyCheck(obj));
+    _Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(mp);
+    _Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(obj);
+
     if (mp->ma_values == NULL || mp->ma_values != _PyObject_InlineValues(obj)) {
         return 0;
     }
     assert(mp->ma_values->embedded == 1);
     assert(mp->ma_values->valid == 1);
     assert(Py_TYPE(obj)->tp_flags & Py_TPFLAGS_INLINE_VALUES);
-    Py_BEGIN_CRITICAL_SECTION(mp);
+
     mp->ma_values = copy_values(mp->ma_values);
```

Comment:
This predates this PR, but I think this should assign to a temporary instead of directly to `mp->ma_values`. Otherwise, we can leave the dict in an invalid state if `copy_values` runs out of memory.